### PR TITLE
Allow plugins to build "pre-configured" token filters

### PR DESCRIPTION
**Pull request by nik9000**
This changes the way we register "pre-built" token filters so that
plugins can declare them and starts to move all of the "pre-built"
token filters out of core. It doesn't finish the job because doing
so would make the change unreviewably large. So this PR includes
a shim that keeps the "old" way of registering "pre-built" token
filters around.

The Lowercase token filter is special because there is a "special"
interaction between it and the lowercase tokenizer. I'm not sure
exactly what to do about it so for now I'm leaving it alone with
the intent of figuring out what to do with it in a followup.

This is a part of #23658

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/indices/analysis/PreBuiltTokenFilters.java [hunk 1]
```diff
@@ -157,73 +69,6 @@ protected boolean isMultiTermAware() {
...
-
-    TRUNCATE(CachingStrategy.ONE) {
-        @Override
-        public TokenStream create(TokenStream tokenStream, Version version) {
-            return new TruncateTokenFilter(tokenStream, 10);
-        }
-    },
-
     // Extended Token Filters
     SNOWBALL(CachingStrategy.ONE) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/indices/analysis/PreBuiltTokenFilters.java]: I stopped migrating here in an attempt to keep the PR reviewable.

File: core/src/main/java/org/elasticsearch/indices/analysis/PreBuiltTokenFilters.java [hunk 2]
```diff
@@ -469,10 +314,16 @@ protected boolean isMultiTermAware() {
...
         cache = PreBuiltCacheFactory.getCache(cachingStrategy);
     }
 
+    public CachingStrategy getCachingStrategy() {
+        return cachingStrategy;
+    }
+
     private interface MultiTermAwareTokenFilterFactory extends TokenFilterFactory, MultiTermAwareComponent {}
 
     public synchronized TokenFilterFactory getTokenFilterFactory(final Version version) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/indices/analysis/PreBuiltTokenFilters.java]: I've preserved this method entirely for `PreBuiltAnalyer.LOWERCASE#getMultiTermComponent`. I'll figure out how to move away from that at some point but I thought that too should wait for another PR.

File: core/src/main/java/org/elasticsearch/index/analysis/PreBuiltTokenFilterFactoryProvider.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    @Override
+    public TokenFilterFactory get(IndexSettings indexSettings, Environment environment, String name, Settings settings) throws IOException {
+        return getTokenFilterFactory(Version.indexCreated(settings));
+    }
+
+    private interface MultiTermAwareTokenFilterFactory extends TokenFilterFactory, MultiTermAwareComponent {}
+
+    private synchronized TokenFilterFactory getTokenFilterFactory(final Version version) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/analysis/PreBuiltTokenFilterFactoryProvider.java]: I've lifted this method pretty much verbatim from `PreBuiltTokenFilters` in an attempt to keep the changes small.

File: core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java [hunk 4]
```diff
@@ -262,6 +271,33 @@ public AnalysisRegistry getAnalysisRegistry() {
         return tokenFilters;
     }
 
+    static Map<String, PreBuiltTokenFilterSpec> setupPreBuiltTokenFilters(List<AnalysisPlugin> plugins) {
+        NamedRegistry<PreBuiltTokenFilterSpec> preBuiltTokenFilters = new NamedRegistry<>("pre built token_filter");
+
+        // Add filters available in lucene-core
+        preBuiltTokenFilters.register("lowercase", new PreBuiltTokenFilterSpec(true, CachingStrategy.LUCENE, (inputs, version) ->
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java]: I debated moving `lowercase` out of core and just keeping standard but there isn't really a technical reason to do that so I decided not to for now.

File: core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java [hunk 5]
```diff
@@ -262,6 +271,33 @@ public AnalysisRegistry getAnalysisRegistry() {
...
+                new LowerCaseFilter(inputs)));
+        preBuiltTokenFilters.register("standard", new PreBuiltTokenFilterSpec(false, CachingStrategy.LUCENE, (inputs, version) ->
+                new StandardFilter(inputs)));
+        /* Note that "stop" is available in lucene-core but it's pre-built version uses a set of English stop words that are in
+         * lucene-analyzers-common so "stop" is defined in the analysis-common module. */
+        
+        // Add token filers declared in PreBuiltTokenFilters until they have all been migrated
+        for (PreBuiltTokenFilters preBuilt : PreBuiltTokenFilters.values()) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java]: This is the shim that I'll drop in a followup.

File: modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/CommonAnalysisFactoryTests.java [hunk 6]
```diff
@@ -46,7 +51,31 @@
...
     protected Map<String, Class<?>> getCharFilters() {
-        Map<String, Class<?>> filters = new HashMap<>(super.getCharFilters());
+        Map<String, Class<?>> filters = new TreeMap<>(super.getCharFilters());
+        return filters;
+    }
+
+    @Override
+    protected Map<String, Class<?>> getPreBuiltTokenFilters() {
+        Map<String, Class<?>> filters = new TreeMap<>(super.getPreBuiltTokenFilters());
+        filters.put("asciifolding", null);
```
Comment (nik9000) [inline modules/analysis-common/src/test/java/org/elasticsearch/analysis/common/CommonAnalysisFactoryTests.java]: I've sort of preserved this `null` behavior from the original test case. It is a bit funky and magical and I'm tempted to remove it but I've kept it for now to get opinions on it.

File: modules/analysis-common/src/test/resources/rest-api-spec/test/analysis-common/40_token_filters.yaml [hunk 7]
```diff
@@ -10,6 +10,18 @@
     - length: { tokens: 1 }
     - match:  { tokens.0.token: Musee d'Orsay }
 
+    - do:
+        indices.analyze:
```
Comment (nik9000) [inline modules/analysis-common/src/test/resources/rest-api-spec/test/analysis-common/40_token_filters.yaml]: I added this when I was debugging around for examples. It isn't strictly needed for this PR but I figure an extra example doesn't hurt.

File: test/framework/src/main/java/org/elasticsearch/indices/analysis/AnalysisFactoryTestCase.java [hunk 8]
```diff
@@ -447,18 +464,24 @@ public void testPreBuiltMultiTermAware() {
...
-            PreBuiltTokenFilters tokenFilter = entry.getKey();
+        Map<String, PreBuiltTokenFilterSpec> preBuiltTokenFilters = AnalysisModule.setupPreBuiltTokenFilters(singletonList(plugin));
+        for (Map.Entry<String, Class<?>> entry : getPreBuiltTokenFilters().entrySet()) {
+            String name = entry.getKey();
             Class<?> luceneFactory = entry.getValue();
             if (luceneFactory == Void.class) {
                 continue;
             }
+            if (luceneFactory == null) {
+                luceneFactory = TokenFilterFactory.lookupClass(toCamelCase(name));
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/indices/analysis/AnalysisFactoryTestCase.java]: This is the other side of that magic `null` behavior that I wasn't super comfortable with. It is convenient though.

**Comment by nik9000** [comment]
As it stands this PR is a net positive in lines of code but I think that is because of the shims that I've left in to keep the PR a manageable size to review. I think it'd be somewhere close to net 0 line difference if it weren't for that.

**Comment by nik9000** [comment]
I talked to @rjernst and he suggested moving the replacing the `PreBuiltTokenFilterSpec` in favor of having the plugin just return the thing that we'll ultimately build with it. We also brainstormed about renaming "pre-built". We decided that preconfigured would be a better name because the components are configured on startup and it more closely reflected the user experience.

**Comment by nik9000** [comment]
@rjernst I think this might be ready for another look.

File: core/src/main/java/org/elasticsearch/index/analysis/PreConfiguredTokenFilter.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        this.create = create;
+    }
+
+    /**
+     * Convenience ctor for token streams that don't vary based on version.
+     */
+    public PreConfiguredTokenFilter(String name, boolean useFilterForMultitermQueries,
+            PreBuiltCacheFactory.CachingStrategy cachingStrategy, Function<TokenStream, TokenStream> create) {
+        this(name, useFilterForMultitermQueries, cachingStrategy, (input, version) -> create.apply(input));
+        // TODO why oh why aren't these all CachingStrategy.ONE? They *can't* vary based on version because they don't get it, right?!
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/PreConfiguredTokenFilter.java]: Sounds correct to me, let's hardcode CachingStrategy.ONE?

Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/analysis/PreConfiguredTokenFilter.java]: I'd been thinking of doing it in a follow. Any objections?

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/PreConfiguredTokenFilter.java]: not at all, this PR is large already

**Comment by nik9000** [comment]
@rjernst can you have another look at this one?

File: core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java [hunk 10]
```diff
@@ -418,28 +419,18 @@ private PrebuiltAnalysis() {
...
                 charFilterFactories.put(name, new PreBuiltCharFilterFactoryFactory(preBuiltCharFilter.getCharFilterFactory(Version.CURRENT)));
             }
             // Char filter aliases
             charFilterFactories.put("htmlStrip", new PreBuiltCharFilterFactoryFactory(PreBuiltCharFilters.HTML_STRIP.getCharFilterFactory(Version.CURRENT)));
+
             this.analyzerProviderFactories = Collections.unmodifiableMap(analyzerProviderFactories);
             this.charFilterFactories = Collections.unmodifiableMap(charFilterFactories);
-            this.tokenFilterFactories = Collections.unmodifiableMap(tokenFilterFactories);
             this.tokenizerFactories = Collections.unmodifiableMap(tokenizerFactories);
+            tokenFilterFactories = preConfiguredTokenFilters;
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: Maybe later (followup, as cleanup after these are all done) these members could be renamed to `preConfigured*`

File: core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java [hunk 11]
```diff
@@ -258,6 +268,40 @@ public AnalysisRegistry getAnalysisRegistry() {
...
+        preConfiguredTokenFilters.register("lowercase",
+                new PreConfiguredTokenFilter("lowercase", true, CachingStrategy.LUCENE, LowerCaseFilter::new));
+        preConfiguredTokenFilters.register("standard",
+                new PreConfiguredTokenFilter("standard", false, CachingStrategy.LUCENE, StandardFilter::new));
+        /* Note that "stop" is available in lucene-core but it's pre-built
+         * version uses a set of English stop words that are in
+         * lucene-analyzers-common so "stop" is defined in the analysis-common
+         * module. */
+        
+        // Add token filers declared in PreBuiltTokenFilters until they have all been migrated
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/indices/analysis/AnalysisModule.java]: typo: filers -> filters

File: core/src/test/java/org/elasticsearch/index/analysis/BuiltInAnalysisFactoryTests.java [hunk 12]
```diff
@@ -19,8 +19,12 @@
 
 package org.elasticsearch.index.analysis;
 
-import org.elasticsearch.AnalysisFactoryTestCase;
+import org.elasticsearch.indices.analysis.AnalysisFactoryTestCase;
+import org.elasticsearch.plugins.AnalysisPlugin;
 
-public class AnalysisFactoryTests extends AnalysisFactoryTestCase {
-    // tests are inherited and nothing needs to be defined here
+public class BuiltInAnalysisFactoryTests extends AnalysisFactoryTestCase {
```
Comment (rjernst) [inline core/src/test/java/org/elasticsearch/index/analysis/BuiltInAnalysisFactoryTests.java]: Can this be PreConfiguredAnalyzerTests? Just a thought, could be later.

**Review by rjernst — APPROVED** [review]
Looks good, thanks for the name change, I think it makes the purpose much clearer (please remember to change the commit message on merge). I left a few suggestions, nothing critical.

**Pull request merged**